src/connectors/cloudtrail.py

```diff
@@ -0,0 +1,211 @@
+    , value:userIdentity.userName::STRING user_identity_username
+    , value:userIdentity.sessionContext.attributes.mfaAuthenticated::STRING user_identity_session_context_attributes_mfa_authenticated
+    , value:userIdentity.sessionContext.attributes.creationDate::STRING user_identity_session_context_attributes_creation_date
+    , value:userIdentity.sessionContext.sessionIssuer.type::STRING user_identity_session_context_session_issuer_type
+    , value:userIdentity.sessionContext.sessionIssuer.principalId::STRING user_identity_session_context_session_issuer_principal_id
+    , value:userIdentity.sessionContext.sessionIssuer.arn::STRING user_identity_session_context_session_issuer_arn
+    , value:userIdentity.sessionContext.sessionIssuer.accountId::STRING user_identity_session_context_session_issuer_account_id
+    , value:userIdentity.sessionContext.sessionIssuer.userName::STRING user_identity_session_context_session_issuer_user_name
+    , value:errorCode::STRING error_code
+    , value:errorMessage::STRING error_message
+    , value:additionalEventData::VARIANT additional_event_data
+    , value:apiVersion::STRING api_version
+    , value:readOnly::BOOLEAN read_only
+    , value:resources::VARIANT resources
+    , value:serviceEventDetails::STRING service_event_details
+    , value:sharedEventID::STRING shared_event_id
+    , value:vpcEndpointID::STRING vpc_endpoint_id
+FROM DATA.{name}_STREAM, table(flatten(input => v:Records))
+WHERE ARRAY_SIZE(v:Records) > 0
+"""
+
+    comment = f"""
+---
+name: {data['name']}
+type: {data['type']}
+stale: {data['stale']}
+method: pipe
+target: {name}_PIPE
+"""
+
+    results = []
+
+    try:
+        db.create_stage(name=name+'_STAGE', url=bucket, prefix=prefix, cloud='aws',
+                        credentials=role, file_format=FILE_FORMAT)
+        results.append({'stage': 'success'})
+    except Exception as e:
+        return e
+
+    try:
+        db.create_table(name=name+'_STAGING', cols={'v': 'variant'})
+        results.append({'staging_table': 'success'})
+    except Exception as e:
+        return e
+
+    try:
+        db.create_stream(name=name+'_STREAM', target=name+'_STAGING')
+        results.append({'stream': 'success'})
+    except Exception as e:
+        return e
+
+    try:
+        pipe_sql = f"COPY INTO DATA.{name}_STAGING FROM @DATA.{name}_STAGE/"
+        db.create_pipe(name=name+'_PIPE', sql=pipe_sql, replace=True)
+        results.append({'pipe': 'success'})
+    except Exception as e:
+        return e
+
+    try:
+        db.create_table(name=name+"_EVENTS_CONNECTION", cols=CLOUDTRAIL_LANDING_TABLE, comment=comment)
+        results.append({'events_table': 'success'})
+    except Exception as e:
+        return e
+
+    try:
+        db.create_task(name=name+'_TASK', schedule='15 minutes',
+                       warehouse=WAREHOUSE, sql=cloudtrail_ingest_task)
+        results.append({'task': 'success'})
+    except Exception as e:
```

Comment:
this is really bad form (as is my auto-handling exceptions in a bunch of places) -- raising an exception means something exceptional happened and returning means everything went as expected